modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextAreaSkin.java

```diff
@@ -571,38 +572,11 @@ private void nextCharacterVisually(boolean moveRight) {
             moveRight = !moveRight;
```

Comment:
This method confuses me as to what it actually wants to do and why
As in, "move right, no, your other right" (!)

isRTL() checks the NodeOrientation so it seems in such a case "right" means "left".
Have you checked the fix with both orientations ?
Seems to me it reverses the meaning of the arrow keys without regard to the text.

And he BreakIterator is getting the next logical position, but this method claims to move to the next visual position.
The method is called nextCharacterVISUALLY, but with this code its actually moving LOGICALLY,
so I'm not seeing how this works for mixed directional text for either setting of "isRTL()" which seems to me to be very questionable to even check here. IMO the arrow key should work the same way no matter whether the UI is laid out with nodes from RTL or LTR.

Perhaps the word "Visually" was not the best choice of name here ? And "right" is questionable too. But I'm not 100% sure.
if you really do want LOGICALLY, then this code will probably work, except I find the isRTL() questionable even then.

What exactly was wrong in the existing code ? Other than it being very complex. How did its calculations go wrong ?

Reply:
PS if I just use simple ASCII text and set Node Orientation to RTL, then with your fix
then left arrow moves right and right arrow moves left.
In JDK 8 left arrow moved left and right arrow moved right.
With an FX 17 build I have around, then the nothing moves.
So it looks like things might have been better when this code was written and some other change broke it.
Your fix unbreaks it, but I'm not entirely sure its restoring intended behaviour.

Reply:
what if you have a mixture of RTL and LTR text in JDK8?
I am going to double check the behavior in JTextArea as well.

Reply:
what is going on??  JTextArea inside a Scene with RTL orientation:

<img width="550" alt="Screenshot 2023-10-09 at 15 23 51" src="https://github.com/openjdk/jfx/assets/107069028/1bd9cda3-71d5-4e41-b475-d5b6c2a010d3">

Reply:
JDK 8u doesn't have the fixes to deal with all of the macOS fonts in current macOS, so its a little hard for me to test Arabic/bidi there since I get garbage text.

I don't see that mirroring that you see, neither setting it on the TextArea or on the Scene, but the text is way off to the right.
Some layout issue.

Reply:
^^ this looks like a separate bug, as it merely mirror images the whole SwingNode or its content.
Setting ComponentOrientation produces desired result, so perhaps we'd need to bind SwingNode's content component orientation to its node orientation.

But anyway, within JTextArea, the RIGHT arrow *always* moves to the right, and the LEFT arrow always moves to the left (of screen), regardless of component orientation which I think might be the "objectively correct" behavior.

Curiously, MS Word does not work this way, instead moving cursor *logically*.

The fixed code works similarly to what MS Word does, reversing the apparent direction within RTL/LTR blocks:
- In the LTR orientation, left moves the cursor left and right - to right within LTR text, which is what we want.
- In the RTL orientation, left moves left and right moves right within RTL text, which is also what we want.

I suppose the decision to work similarly to MS Word (and differently from swing?) was made early on and we are stuck with it.

Sooo, I guess the fix works as designed?
(I agree with your other objections to naming and javadoc, will fix).

Reply:
> JDK 8u doesn't have the fixes to deal with all of the macOS fonts in current macOS, so its a little hard for me to test Arabic/bidi there since I get garbage text.

if I remember correctly, explicitly setting an Arabic font removes the garbage.

> I don't see that mirroring that you see, neither setting it on the TextArea or on the Scene, but the text is way off to the right.

I've added this to the MonkeyTester (standalone revision)
https://github.com/andy-goryachev-oracle/MonkeyTest
Tools -> JTextArea 
and
Window -> Orientation: RTL